lib/parsers/json_parser/validator.rb

```diff
@@ -0,0 +1,318 @@
+##
+# This class checks whether a given file is a valid JSON file.
+# The validation process DOES NOT assemble an object with the contents of the JSON file in memory,
+# Instead, it implements a simple state-machine-like that digests the contents of the file while traversing
+# the hierarchy of nodes in the document.
+#
+# Although this is based on the IETF standard (https://www.rfc-editor.org/rfc/rfc8259),
+# it does cut a few corners for the sake of simplicity. For instance, instead of validating
+# Numbers, "true", "false" and "null" tokens, it supports a type called Literal to hold generic sequences of characters.
+# This decision makes the implementation simpler while being a good-enough approach to identify JSON files.
+#
+# There is also a cap. Large files are not read all the way through. Instead, if the beginning of file is
+# JSON-compliant, it is assumed that the file is a JSON file.
+
+class FormatParser::JSONParser::Validator
+  class JSONParserError < StandardError
+  end
+
+  MAX_SAMPLE_SIZE = 1024
+  MAX_LITERAL_SIZE = 30 # much larger then necessary.
+  ESCAPE_CHAR = "\\"
+  WHITESPACE_CHARS = [" ", "\t", "\n", "\r"]
+  LITERALS_CHAR_TEMPLATE = /\w|[+\-.]/ # any alphanumeric, "+", "-" and "."
+
+  def initialize(io)
+    @io = io
+    @current_node = nil # :object, :array, :string, :literal
+    @parent_nodes = []
+    @current_state = :awaiting_root_node
+    @escape_next = false
+    @current_literal_size = 0
+    @pos = 0
+
+    @all_parsers = {}
+
+    @execution_stats = {
+      array: 0,
+      object: 0,
+      literal: 0,
+      string: 0
+    }
+
+    setup_transitions
+  end
+
+  def validate
+    char_reader = FormatParser::UTF8Reader.new(@io)
+
+    while (c = char_reader.read_char)
+      @pos += 1
+      parse_char c
+
+      # Halt validation if the sampling limit is reached.
+      if @pos >= MAX_SAMPLE_SIZE
+        raise JSONParserError, "Invalid JSON file" if @current_state == :awaiting_root_node
+        return false
+      end
+    end
+
+    # Raising error in case the EOF is reached earlier than expected
+    raise JSONParserError, "Incomplete JSON file" if @current_state != :closed
+    true
+  rescue  FormatParser::UTF8Reader::UTF8CharReaderError
+    raise JSONParserError, "Invalid UTF-8 character"
+  end
+
+  def stats(node_type)
+    @execution_stats[node_type]
+  end
+
+  private
+
+  def setup_transitions
+    when_its :awaiting_root_node, ->(c) {
+      read_whitespace(c) or
+        start_object(c) or
```

Comment:
Curious whether there's any reason for the extra indentation here 🤔

Reply:
Rubocop requires the additional indentation for expressions spawning multiple lines.

Reply:
Interesting!